### PR TITLE
Replacing handlebars for lodash template

### PR DESCRIPTION
**Pull request by nuxlli**
This PR is a suggestion to solve #3.

- Keep support for templates;
- Adds support to customize interpolation regex
- Replaces the handlebars for lodash template;

The lodash template is more simple (and smaller) than handlerbars.

### CODE REVIEW

**Comment by macrozone** [comment]
@nuxlli: is it possible to install your version of message-box before the merge? 

i tried npm install message-box@https://github.com/nuxlli/node-message-box.git#46dbc90b822feb4785b6d6ec5cd0287d6d1ad152 --save but it installs an empty package 

edit: and thank you for your efforts! hope this gets merged soon!

**Comment by nuxlli** [comment]
@macrozone: Unfortunately it is not possible. This project uses babel, so a build phase is necessary before using it, but there is a limitation on installing packages that need this phase from their git repositories, this is a complicated problem and involves a long discussion around the npm: https://github.com/npm/npm/issues/3055

To help you, I've temporarily made a branch, already with the build done and with the `dist/MessageBox.js` file, you can install using:

`npm install message-box@https://github.com/nuxlli/node-message-box.git#replacing-handlerbars-with-build`

I promise not to remove the repository and break some of its installation until the merge of this PR ;)

**Comment by macrozone** [comment]
@nuxlli thx, will try that!

**Comment by macrozone** [comment]
@nuxlli on server i got an error:

```
Exception in queued task: SyntaxError: Unexpected token )
   at Function (native)
   at /Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/lodash.template/index.js:1089:12
   at apply (/Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/lodash.template/index.js:74:25)
   at /Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/lodash.template/index.js:1126:12
   at apply (/Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/lodash.template/index.js:76:25)
   at /Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/lodash.template/index.js:317:12
   at template (/Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/lodash.template/index.js:1088:16)
   at MessageBox.message (/Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/message-box/dist/MessageBox.js:97:71)
   at SimpleSchema.messageForError (/Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/simpl-schema/dist/SimpleSchema.js:716:30)
   at ValidationContext.keyErrorMessage (/Users/macrozone/git/panter/pvl/velo-mittwoch/app/node_modules/simpl-schema/dist/ValidationContext.js:194:33)
```

I dumped the code that it wants to eval():

```
function(obj) {
obj || (obj = {});
var __t, __p = '';
with (obj) {
__p +=
((__t = ({label)) == null ? '' : __t) +
'} must be of type ' +
((__t = (dataType)) == null ? '' : __t);
}
return __p
}
```

the error seems to be in this line: `((__t = ({label)) == null ? '' : __t) +`

where the brackets seems to be wrong.

The coresponding string is: `expectedType: '{{{label}}} must be of type {{dataType}}',`


so tripple brackets case this error. Do you know how to fix this?

**Comment by macrozone** [comment]
@nuxlli I changed the default regex to be more forgiving: /{{{?([\s\S]+?)}?}}/g;

its not 100% correct but works for tripple and double brackets (and also for mixed versions...)

see https://github.com/panter/node-message-box/commit/e0312ee70c6d8f92f2bc3f951a64de185b0d5cc0

**Comment by nuxlli** [comment]
@macrozone thanks for the feedback.

In fact, your version does not support cases with more braces. I have provided a somewhat more robust version and have already submitted a correcting commit.

You can test with: https://regex101.com/r/8sRC8b/2/

Note: I have also made the branch rebase with the compiled script for your convenience ;)

**Comment by macrozone** [comment]
hi @nuxlli thx very much!

With your regex, `{{{label}}}` would keep the most outer braces, wouldn't it? 

E.g. `{{{label}}} must be a string` (with label="foo") would result in `{foo} must be a string` 

Problem is that SimpleSchema defines the default messages with tripple braces.

Edit: in your tests you even do confirm this. But i think it should remove all three braces

**Comment by nuxlli** [comment]
@macrozone, had not seen SimpleSchema make use of this handlerbars option.

What do you think of this solution https://regex101.com/r/8sRC8b/4 ?

She has a little glitch but I think it's acceptable for the hour.

**Comment by aldeed** [comment]
@nuxlli @macrozone Thanks! I can merge this when you're confident in it. I don't have much time to look closely at the moment, but a couple points:

- SS uses triple brace because labels can be HTML strings. So it might be good to add a test for making sure HTML comes through without being escaped. Also, double brace should be "escape" delimiter maybe?
- I'm not sure if things like {{#if}} are currently used by people. I believe they'd be able to use "evaluate" delimiter instead, but it will probably be safest to consider this a breaking change.

**Comment by nuxlli** [comment]
@aldeed thanks for the feedback.

It was a bit more complicated than my initial analysis :)

But I believe I have arrived in a functional version:

- Setted the "escape" configuration of the lodash. Now two brace do not make automatic escape and three do so (like a handlebars). Also added tests;

- With the "evaluate" option made it possible to use `if`, but in a format different from that of handlebars, but this should work: `'{{# if (value) { }}true{{# } else { }}false{{# } }}'`

File: lib/MessageBox.js
Comment (aldeed) [inline lib/MessageBox.js]: Can you add the same lines for `evaluate` and `escape` here?

File: lib/MessageBox.js
Comment (aldeed) [inline lib/MessageBox.js]: Rather than adding the default expression here, can you set initial values for the static props? So something like

```js
this.interpolate = interpolate || MessageBox.interpolate;
this.evaluate = evaluate || MessageBox.evaluate;
this.escape = escape || MessageBox.escape;
```

And after the constructor function:

```js
static interpolate = /{{{([^\{\}#][\s\S]+?)}}}/g;
static escape = /{{([^\{\}#][\s\S]+?)}}/g;
```

**Review by aldeed — CHANGES_REQUESTED** [review]

**Comment by aldeed** [comment]
@nuxlli See a couple code comments. Also, I think we should not have the default "evaluate" expression. Allow them to set one but don't set an initial `static evaluate`. I don't necessarily want to encourage crazy logic in messages, and there could be security issues with evaluating something unexpectedly.

Instead maybe you can add your example regex and syntax to the readme. In general, the options should be documented in the readme. Thanks!

**Comment by nuxlli** [comment]
@aldeed I agree with you on all points, I made the changes.

**Comment by macrozone** [comment]
@nuxlli could you create a installable branch? :-)

**Comment by nuxlli** [comment]
@macrozone done, sorry about that, I completely forgot

File: lib/MessageBox.js
Comment (nuxlli) [reply lib/MessageBox.js]: I preferred to use constants, but it was done.

File: lib/MessageBox.js
Comment (nuxlli) [reply lib/MessageBox.js]: I really forgot that detail.

**Comment by macrozone** [comment]
@nuxlli no worries, let's hope @aldeed takes a look at this and can release a new version soon ;-)

**Review by aldeed — APPROVED** [review]

**Pull request merged**
